workflow/controller/operator.go

```diff
@@ -181,6 +181,49 @@ func newWorkflowOperationCtx(wf *wfv1.Workflow, wfc *WorkflowController) *wfOper
 	return &woc
 }
 
+// Determines if this workflow is definitely after the introduction of
+// the NodeFlag field.
+// used to avoid the O(n*m) loop in maybeUpgradeWithNodeFlags
+func (woc *wfOperationCtx) definitePostNodeFlagsWf() bool {
+	for _, node := range woc.wf.Status.Nodes {
+		if node.NodeFlag != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (woc *wfOperationCtx) maybeUpgradeWithNodeFlags() {
```

Comment:
Should we force add a non nil nodeFlag in this to the root node to ensure `definitePostNodeFlagsWf` returns true as fast as possible?

Should new workflows created by this controller also do that at creation?

Reply:
I don't think we need to, even for extremely large workflows (100,000 nodes) , a simple for loop over their nodes shouldn't take much time. I think there are other parts of the codebase that will suffer before these two functions become noticeably slow.

Reply:
Ok, I'll accept that.